imagescan/subscriber.go

```diff
@@ -290,41 +301,98 @@ func (s *Subscriber) getScanMode(img *image) string {
 	return mode
 }
 
-func (s *Subscriber) sendImageOwnerChanges(ctx context.Context, images []*image) {
-	var errg errgroup.Group
-	errg.SetLimit(10)
-
+func (s *Subscriber) sendImagesResourcesChanges(ctx context.Context) {
+	images := s.delta.getImages()
+	var imagesChanges []castai.Image
 	for _, img := range images {
-		img := img
 		if img.ownerChanges.empty() {
 			continue
 		}
+		imagesChanges = append(imagesChanges, castai.Image{
+			ID:           img.id,
+			Architecture: img.architecture,
+			ResourcesChange: castai.ResourcesChange{
+				ResourceIDs:        img.ownerChanges.addedIDS,
+				RemovedResourceIDs: img.ownerChanges.removedIDs,
+			},
+		})
+	}
+	if len(imagesChanges) == 0 {
+		return
+	}
 
-		select {
-		case <-ctx.Done():
-			return
-		default:
+	s.log.Info("sending images resources changes")
+	report := &castai.ImagesResourcesChange{
+		Images: imagesChanges,
+	}
+	if err := s.client.SendImagesResourcesChange(ctx, report); err != nil {
+		s.log.Errorf("sending images resources changes: %v", err)
+		return
+	}
+
+	// Clear changes state.
+	for _, img := range images {
+		if img.ownerChanges.empty() {
+			continue
 		}
+		img.ownerChanges.clear()
+	}
+}
 
-		errg.Go(func() error {
-			ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
-			defer cancel()
-			if err := s.client.SendImageMetadata(ctx, &castai.ImageMetadata{
-				ImageName:          img.name,
-				ImageID:            img.id,
-				Architecture:       img.architecture,
-				ResourceIDs:        img.ownerChanges.addedIDS,
-				RemovedResourceIDs: img.ownerChanges.removedIDs,
-			}); err != nil {
-				return err
-			}
-			// Reset owner changes state.
-			img.ownerChanges.clear()
-			return nil
+func (s *Subscriber) sendFullSnapshotImageResources(ctx context.Context) error {
+	s.log.Info("sending initial full images resources changes")
+	images := s.delta.getImages()
+	report := &castai.ImagesResourcesChange{
+		FullSnapshot: true,
+	}
+	for _, img := range images {
+		report.Images = append(report.Images, castai.Image{
+			ID:           img.id,
+			Architecture: img.architecture,
+			ResourcesChange: castai.ResourcesChange{
+				ResourceIDs: lo.Keys(img.owners),
+			},
 		})
 	}
+	return s.client.SendImagesResourcesChange(ctx, report)
+}
+
+func (s *Subscriber) syncFromRemoteState(ctx context.Context) {
+	images := s.delta.getImages()
+	now := s.timeGetter().UTC()
+	imagesWithNotSyncedState := lo.Filter(images, func(item *image, index int) bool {
+		return !item.scanned && item.lastRemoteSyncAt.Before(now.Add(-10*time.Minute))
+	})
+
+	if len(imagesWithNotSyncedState) == 0 {
+		return
+	}
+
+	imagesIds := lo.Map(imagesWithNotSyncedState, func(item *image, index int) string {
+		return item.id
+	})
+	s.log.Debugf("sync images state from remote")
```

Comment:
debug

Reply:
Left it on purpose. This should not log much.